### PR TITLE
Feature/job executor

### LINKED ISSUE #2: Implement Workflow Analyzer Module
# Issue 2: Workflow Analyzer Module

## Overview

| Field | Value |
|-------|-------|
| **Issue Title** | Workflow Analyzer Module |
| **Phase** | 1 - Core Infrastructure |
| **Priority** | High |
| **Dependencies** | Issue 1 (CLI Framework) |
| **Blocked By** | Issue 1 |
| **Blocks** | Issues 3, 5, 6, 12 |

## Description

Parse GitHub Actions YAML workflow files to extract jobs, matrix configurations, dependencies, and environment requirements. This module is the intelligence layer that translates workflow definitions into structured data consumed by the executor, image matcher, and orchestrator.

## Detailed Doc: [Issue 2 - Workflow Analyzer](https://github.com/CppDigest/team-brain/blob/ff5a1e13becbf0bbb10b6b2cc2283037979ee527/2026-02/2026-02-06/brad/Local%20CI%20Phase%201/Issue%202%20-%20Workflow%20Analyzer%20Module.md)

The analyzer must handle:
- Complex matrix strategies with `include` entries
- Container-based and runner-based jobs
- Job dependencies (`needs`)
- Conditional execution (`if`)
- Environment variable resolution
- Step-level analysis (identify build system: B2 vs CMake)
- Multiple workflow files

## Goals

1. **Accurate Parsing**: Correctly extract all fields from capy's ci.yml
2. **Structured Output**: Produce well-typed data models for downstream consumers
3. **Platform Classification**: Automatically categorize matrix entries by platform
4. **Requirement Extraction**: Identify OS, compiler, packages, and build tools needed
5. **Extensibility**: Support any Boost library's CI workflow, not just capy

---

### Reference: `cppalliance/capy/.github/workflows/ci.yml`

```yaml
# Key structural elements to parse:

# 1. Trigger events
on:
  pull_request:
  push:
    branches: [master, develop, bugfix/**, feature/**, fix/**, pr/**]

# 2. Concurrency
concurrency:
  group: ${{format('{0}:{1}', github.repository, github.ref)}}
  cancel-in-progress: true

# 3. Global environment
env:
  GIT_FETCH_JOBS: 8
  NET_RETRY_COUNT: 5
  DEFAULT_BUILD_VARIANT: debug,release
  UBSAN_OPTIONS: "print_stacktrace=1"
  DEBIAN_FRONTEND: "noninteractive"
  TZ: "Europe/London"

# 4. Matrix strategy (inside jobs.build)
strategy:
  fail-fast: false
  matrix:
    include:
      - compiler: "gcc"
        version: "15"
        cxxstd: "20"
        runs-on: "ubuntu-latest"
        container: "ubuntu:25.04"
        b2-toolset: "gcc"
        name: "GCC 15: C++20"
        shared: true
        build-type: "Release"
        build-cmake: true
        # ...

# 5. Steps (build systems)
steps:
  - uses: alandefreitas/cpp-actions/b2-workflow@v1.9.0   # B2 build
  - uses: alandefreitas/cpp-actions/cmake-workflow@v1.9.0 # CMake build
```

### Matrix Entry Fields (Complete)

| Field | Type | Example | Purpose |
|-------|------|---------|---------|
| `compiler` | string | `gcc`, `clang`, `msvc`, `mingw`, `apple-clang` | Compiler family |
| `version` | string | `15`, `20`, `14.42`, `*` | Compiler version |
| `cxxstd` | string | `20`, `20,23` | C++ standard(s) |
| `latest-cxxstd` | string | `20`, `23` | Latest standard for CMake |
| `cxx` | string | `g++-15`, `clang++-20` | C++ compiler binary |
| `cc` | string | `gcc-15`, `clang-20` | C compiler binary |
| `runs-on` | string | `ubuntu-latest`, `windows-2022`, `macos-26` | GitHub runner |
| `container` | string | `ubuntu:25.04`, `ubuntu:24.04` | Docker container image |
| `b2-toolset` | string | `gcc`, `clang`, `msvc-14.4` | B2 toolset name |
| `generator` | string | `Visual Studio 17 2022`, `MinGW Makefiles` | CMake generator |
| `is-latest` | bool | `true` | Whether latest compiler |
| `is-earliest` | bool | `true` | Whether earliest compiler |
| `name` | string | `GCC 15: C++20` | Human-readable name |
| `shared` | bool | `true`/`false` | Build shared libraries |
| `build-type` | string | `Release`, `Debug`, `RelWithDebInfo` | Build configuration |
| `build-cmake` | bool | `true` | Also run CMake build |
| `asan` | bool | `true` | Enable AddressSanitizer |
| `ubsan` | bool | `true` | Enable UBSanitizer |
| `coverage` | bool | `true` | Enable coverage |
| `x86` | bool | `true` | Build for 32-bit |
| `install` | string | `lcov wget unzip` | Additional packages |
| `cxxflags` | string | `--coverage -fprofile-arcs` | Extra C++ flags |
| `ccflags` | string | `--coverage -fprofile-arcs` | Extra C flags |

### Capy's 14 Matrix Entries

```
 #  │ Name                          │ Compiler    │ OS              │ Container      │ Build Systems │ Variants
────┼───────────────────────────────┼─────────────┼─────────────────┼────────────────┼───────────────┼──────────
  1 │ MSVC 14.42: C++20             │ msvc 14.42  │ windows-2022    │ -              │ B2 + CMake    │ -
  2 │ MSVC 14.34: C++20 (shared)    │ msvc 14.34  │ windows-2022    │ -              │ B2            │ shared
  3 │ MinGW: C++20                  │ mingw *     │ windows-2022    │ -              │ B2 + CMake    │ -
  4 │ Apple-Clang (asan+ubsan)      │ apple-clang │ macos-26        │ -              │ B2            │ asan, ubsan
  5 │ GCC 15: C++20                 │ gcc 15      │ ubuntu-latest   │ ubuntu:25.04   │ B2 + CMake    │ shared
  6 │ GCC 15: C++20 (asan+ubsan)    │ gcc 15      │ ubuntu-latest   │ ubuntu:25.04   │ B2            │ asan, ubsan
  7 │ GCC 12: C++20                 │ gcc 12      │ ubuntu-latest   │ ubuntu:22.04   │ B2            │ shared
  8 │ GCC 13: C++20 (coverage)      │ gcc 13      │ ubuntu-24.04    │ -              │ CMake         │ coverage
  9 │ Clang 20: C++20-23            │ clang 20    │ ubuntu-latest   │ ubuntu:24.04   │ B2 + CMake    │ shared
 10 │ Clang 20: C++20 (asan+ubsan)  │ clang 20    │ ubuntu-latest   │ ubuntu:24.04   │ B2            │ asan, ubsan
 11 │ Clang 17: C++20               │ clang 17    │ ubuntu-24.04    │ -              │ B2            │ -
 12 │ Clang 20: C++20-23 (x86)      │ clang 20    │ ubuntu-latest   │ ubuntu:24.04   │ B2            │ x86
```

### Capy's CI Steps (in order)

```
Step 1: Clone Boost.Capy      → actions/checkout@v4
Step 2: Setup C++              → alandefreitas/cpp-actions/setup-cpp@v1.9.0
Step 3: Install packages       → alandefreitas/cpp-actions/package-install@v1.9.0
Step 4: Clone Boost            → alandefreitas/cpp-actions/boost-clone@v1.9.0
Step 5: ASLR Fix               → sysctl vm.mmap_rnd_bits=28
Step 6: Patch Boost             → shell script (copy module into boost tree)
Step 7: Boost B2 Workflow       → alandefreitas/cpp-actions/b2-workflow@v1.9.0
Step 8: CMake Workflow          → alandefreitas/cpp-actions/cmake-workflow@v1.9.0
Step 9: FlameGraph              → alandefreitas/cpp-actions/flamegraph@v1.9.0
Step 10: Codecov                → shell script (lcov + codecov upload)
```

### Functional

- [ ] Parse capy's `ci.yml` and extract all 14 matrix entries correctly
- [ ] Classify all entries by platform (10 Linux, 3 Windows, 1 macOS)
- [ ] Detect compiler family and version for all entries
- [ ] Detect container image for containerized entries
- [ ] Detect build system (B2, CMake, or both) per entry
- [ ] Extract package requirements
- [ ] Parse job dependencies
- [ ] Filter by platform, compiler, variant
- [ ] Search by name pattern
- [ ] JSON output matches expected schema
- [ ] Handle multiple workflow files

### Non-Functional

- [ ] Full parse completes in < 2 seconds
- [ ] Clear error messages for malformed YAML
- [ ] Works on Windows (PowerShell), Linux, macOS
- [ ] No yq version-specific issues (mikefarah/yq v4+)

### Testing

- [ ] Unit tests for all models
- [ ] Unit tests for yq wrapper
- [ ] Unit tests for analyzer with fixtures
- [ ] Integration test with real capy ci.yml
- [ ] Edge case coverage (empty matrix, no container, wildcard version)
- [ ] Test coverage > 85%

---

## Acceptance Criteria

1. `localci analyze .github/workflows/ci.yml` shows all 14 entries correctly
2. `localci analyze --format json` produces valid parseable JSON
3. `localci list --platform linux` shows exactly 10 entries
4. `localci list --compiler gcc` shows exactly 4 GCC entries
5. Each entry has correct compiler, platform, container, and variant classification
6. Job dependency ordering is correct
7. Running against any Boost library's CI workflow produces sensible output

---

## References

- [yq Documentation](https://mikefarah.gitbook.io/yq/)
- [GitHub Actions Workflow Syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions)
- [capy CI Workflow](https://github.com/cppalliance/capy/blob/develop/.github/workflows/ci.yml)
- [cpp-actions](https://github.com/alandefreitas/cpp-actions) - Actions used by capy

### PR DESCRIPTION
**Pull request by bradjin8**
Resolve [Issue 5: Job Executor with act Integration](https://github.com/CppDigest/local-ci-test-system/issues/3)

## Summary
Execute GitHub Actions jobs locally using act (nektos/act), with proper image loading, command construction, output capture, result parsing, and container cleanup. This is the execution engine that turns parsed workflow data into actual test runs.

## Goals
- Reliable Execution: Run any capy matrix entry locally with correct results
- Efficient Image Handling: Load pre-built images, avoid pulling from registry
- Observable Output: Stream logs in real-time, capture for later review
- Graceful Failure: Handle timeouts, crashes, and resource exhaustion
- Minimal Overhead: Executor adds < 5 seconds to job runtime

## Summary by CodeRabbit

* **New Features**
  * Fully implemented CLI: analyze, list, logs, run, and status commands with JSON/table/YAML output, filtering, dry-run, and file persistence
  * Local execution engine with job summaries, log capture, and colored status reports
  * Workflow analysis and serialization plus Docker/image utilities and improved config path expansion

* **Tests**
  * Expanded test suite covering workflow analysis, execution, serialization, Docker, and CLI end-to-end scenarios

### CODE REVIEW

**Review by AuraMindNest — COMMENTED** [review]
### 1. `--execution-id` Cannot Work
**Files:** `cli/run.py`, `cli/status.py`, `cli/logs.py`
- **Problem:** `localci run` generates `execution_id` but only saves to `last-run.json`. Options `status --execution-id X` and `logs -e X` look for `logs_dir/X.json` which is never created.

### 2. Path with `~` Not Expanded Everywhere
**Files:** `core/config.py`, `cli/run.py:314`, `cli/status.py:52-58`, `cli/logs.py:71-77`
- **Problem:** Config path `logging: directory: ~/.localci/logs` → `JobExecutor` expands it, but `run.py` writes summary to unexpanded path, `status`/`logs` read from unexpanded path.

### 3. `analyze --event` Has No Effect
**Files:** `cli/analyze.py:153`, `core/workflow.py`
- **Problem:** CLI accepts `--event` and passes to `analyzer.analyze(path, event)` but parameter is never used.

### 4. `list --enabled / --disabled` Ignored
**File:** `cli/list.py`
- **Problem:** Options defined but no filtering by `config.jobs.include/exclude`.

**Review by AuraMindNest — APPROVED** [review]

**Comment by wpak-ai** [comment]
@coderabbitai review this PR

**Comment by bradjin8** [comment]
@coderabbitai review

**Review by wpak-ai — APPROVED** [review]
*Please address all review comments from CodeRabbit before merge*, otherwise looks good to me.
Minor note: The current PR is too tailored for `capy`. We want a more flexible solution. Since this is a prototype, let's plan to modify it later.

File: cli/localci/cli/logs.py [hunk 1]
```diff
@@ -38,6 +67,109 @@ def logs(
...
+        lines = lines[-tail:]
+
+    content = "\n".join(lines)
+
+    # Save to file
+    if output_file:
+        out = Path(output_file)
+        out.write_text(content, encoding="utf-8")
+        print_info(f"Logs saved to {out}")
+        return
```
[the remark this answers is not in the document]
Comment (bradjin8) [reply cli/localci/cli/logs.py:139]: Updates in cli/localci/cli/logs.py (lines 129–137):
Create parent directories – Before writing, call out.parent.mkdir(parents=True, exist_ok=True) so the full path exists and FileNotFoundError is avoided when the path has missing directories (e.g. --output /tmp/new-dir/out.txt).
Guard the write with try/except – Wrap the mkdir and write_text in try/except (FileNotFoundError, OSError). OSError is used instead of IOError (Python 3 makes IOError an alias of OSError) so both “file not found” and other I/O errors are handled.
User-visible error – On exception, call print_error(f"Failed to write logs to {out}: {exc}") and then ctx.exit(1) so the user sees a clear message and the command exits with a non-zero status.

File: cli/localci/cli/status.py [hunk 2]
```diff
@@ -28,9 +48,82 @@ def status(
...
+    # Table output
+    console.print()
+    print_info(f"Execution: {summary.execution_id}")
+    console.print()
+
+    table = make_table(
+        "#", "Name", "Status", "Duration", "Image",
+        title="Job Results",
+    )
+    for r in sorted(summary.results, key=lambda x: x.matrix_index):
```
[the remark this answers is not in the document]
Comment (bradjin8) [reply cli/localci/cli/status.py]: Used a sort key that handles None by ordering by a tuple:
(x.matrix_index is None, x.matrix_index).

File: cli/localci/core/config.py [hunk 3]
```diff
@@ -135,6 +157,12 @@ class LocalCIConfig(BaseModel):
     version: int = 1
     workflow: Path = Field(default=Path(".github/workflows/ci.yml"))
     event: str = "push"
+
+    @field_validator("workflow", mode="after")
+    @classmethod
+    def expand_workflow(cls, v: Path) -> Path:
+        return _expand_path(v)
```
[the remark this answers is not in the document]
Comment (bradjin8) [reply cli/localci/core/config.py:167]: Set validate_default=True on the workflow Field so the validator runs for the default as well
